Maple2.Server.Game/Manager/ShopManager.cs

```diff
@@ -0,0 +1,609 @@
+                if (!buyData.TimeRanges.Any(time => time.StartTimeOfDay > totalSeconds || time.EndTimeOfDay < totalSeconds)) {
+                    session.Send(ShopPacket.Error(ShopError.s_err_invalid_item_cannot_buy_by_period));
+                    return;
+                }
+            }
+
+            if (buyData.Days.Count > 0) {
+                if (!buyData.Days.Contains(ToShopBuyDay(DateTime.Now.DayOfWeek))) {
+                    session.Send(ShopPacket.Error(ShopError.s_err_invalid_item_cannot_buy_by_period));
+                    return;
+                }
+            }
+        }
+
+        if (shopItem.StockCount > 0 && shopItem.StockPurchased + quantity > shopItem.StockCount) {
+            session.Send(ShopPacket.Error(ShopError.s_err_lack_shopitem));
+            return;
+        }
+
+        if (shopItem.RequireAchievementId > 0 && !session.Achievement.HasAchievement(shopItem.RequireAchievementId, shopItem.RequireAchievementRank)) {
+            return;
+        }
+
+        if (shopItem.RequireGuildTrophy > 0 && session.Guild.Guild != null && session.Guild.Guild.AchievementInfo.Total < shopItem.RequireGuildTrophy) {
+            session.Send(ShopPacket.Error(ShopError.s_err_lack_guild_trophy));
+            return;
+        }
+
+        // TODO: Guild Merchant Type/Level, Championship, Alliance
+
+        Item item = shopItem.Item.Clone();
+        item.Amount = quantity;
+        if (!session.Item.Inventory.CanAdd(item)) {
+            session.Send(ShopPacket.Error(ShopError.s_err_inventory));
+            return;
+        }
+
+        int price = shopItem.Cost.SaleAmount > 0 ? shopItem.Cost.SaleAmount * quantity : shopItem.Cost.Amount * quantity;
+        if (!Pay(shopItem.Cost, price)) {
+            return;
+        }
+
+        if (activeShop.RestockData != null && shopItem.StockCount > 0) {
+            UpdateStockCount(activeShop.Id, shopItem, quantity);
+            session.Send(ShopPacket.Update(shopItem.Id, quantity * shopItem.StockPurchased));
+        }
+
+        session.Item.Inventory.Add(item, true);
+        session.Send(ShopPacket.Buy(shopItem, quantity, price));
+    }
+
+    public void PurchaseBuyBack(int id) {
+        if (activeShop == null) {
+            return;
+        }
+
+        if (!buyBackItems.TryGetValue(id, out BuyBackItem? buyBackItem)) {
+            session.Send(ShopPacket.Error(ShopError.s_err_invalid_item));
+            return;
+        }
+
+        if (!session.Item.Inventory.CanAdd(buyBackItem.Item)) {
+            session.Send(ShopPacket.Error(ShopError.s_err_inventory));
+            return;
+        }
+
+        if (!Pay(new ShopCost {
+                Type = ShopCurrencyType.Meso,
+                Amount = (int) buyBackItem.Price,
+            }, (int) buyBackItem.Price)) {
+            return;
+        }
+
+        session.Item.Inventory.Add(buyBackItem.Item, true);
+        buyBackItems.Remove(id);
```

Comment:
I would generally prefer removing before adding to avoid exploits (and the remove should succeed)